duperemove.8
.TH "duperemove" "8" "September 2016" "Version 0.11"
.SH "NAME"
duperemove \- Find duplicate extents and print them to stdout
.SH "SYNOPSIS"
\fBduperemove\fR \fI[options]\fR \fIfiles...\fI
.SH "DESCRIPTION"
.PP
\fBduperemove\fR is a simple tool for finding duplicated extents and
submitting them for deduplication. When given a list of files it will
hash the contents of their extents and compare those hashes to each
other, finding and categorizing extents that match each other. When
given the \fB-d\fR option, \fBduperemove\fR will submit those extents
for deduplication using the Linux kernel extent-same ioctl.

\fBduperemove\fR can store the hashes it computes in a \fIhashfile\fR. If
given an existing hashfile, \fBduperemove\fR will only compute hashes
for those files which have changed since the last run.  Thus you can run
\fBduperemove\fR repeatedly on your data as it changes, without having to
re-checksum unchanged data.  For more on hashfiles see the
\fB--hashfile\fR option below as well as the \fIExamples\fR section.

\fBduperemove\fR can also take input from the \fBfdupes\fR program, see the
\fB--fdupes\fR option below.

.SH "GENERAL"
Duperemove has two major modes of operation one of which is a subset
of the other.

.SS "Readonly / Non-deduplicating Mode"

When run without \fB-d\fR (the default) duperemove will print out one or
more tables of matching extents it has determined would be ideal
candidates for deduplication. As a result, readonly mode is useful for
seeing what duperemove might do when run with \fB-d\fR.

Generally, duperemove does not concern itself with the underlying
representation of the extents it processes. Some of them could be
compressed, undergoing I/O, or even have already been deduplicated. In
dedupe mode, the kernel handles those details and therefore we try not
to replicate that work.

.SS "Deduping Mode"

This functions similarly to readonly mode with the exception that the
duplicated extents found in our "read, hash, and compare" step will
actually be submitted for deduplication. Extents that have already
been deduped will be skipped. An estimate of the total data
deduplicated will be printed after the operation is complete. This
estimate is calculated by comparing the total amount of shared bytes
in each file before and after the dedupe.

.SH "OPTIONS"
\fIfiles\fR can refer to a list of regular files and directories or be
a hyphen (-) to read them from standard input.
If a directory is specified, all regular files within it will also be
scanned. Duperemove can also be told to recursively scan directories with
the '-r' switch.

.TP
\fB\-r\fR
Enable recursive dir traversal.

.TP
\fB\-d\fR
De-dupe the results - only works on \fIbtrfs\fR and \fIxfs\FR.

.TP
\fB\-A\fR
Opens files readonly when deduping; currently requires root privileges
(and is enabled by default for root). Allows use on readonly snapshots
or when the file might be open for exec.

.TP
\fB\-h\fR
Print numbers in human-readable format.

.TP
\fB\-q\fR
Quiet mode. Duperemove will only print errors and a short summary of any dedupe.

.TP
\fB\--hashfile=hashfile\fR
Use a file for storage of hashes instead of memory. This option drastically
reduces the memory footprint of duperemove and is recommended when your data
set is more than a few files large.  \fIHashfiles\fR are also reusable,
allowing you to further reduce the amount of hashing done on subsequent
dedupe runs.

If \fIhashfile\fR does not exist it will be created.  If it exists,
\fBduperemove\fR will check the file paths stored inside of it for changes.
Files which have changed will be rescanned and their updated hashes will be
written to the \fIhashfile\fR.  Deleted files will be removed from the \fIhashfile\fR.

New files are only added to the \fIhashfile\fR if they are discoverable via
the \fIfiles\fR argument.  For that reason you probably want to provide the
same \fIfiles\fR list and \fB-r\fR arguments on each run of
\fBduperemove\fR.  The file discovery algorithm is efficient and will only
visit each file once, even if it is already in the \fIhashfile\fR.

Adding a new path to a hashfile is as simple as adding it to the \fIfiles\fR
argument.

When deduping from a hashfile, duperemove will avoid deduping files which
have not changed since the last dedupe.

.TP
\fB\-L\fR
Print all files in the hashfile and exit. Requires the \fB\--hashfile\fR option.
Will print additional information about each file when run with \fB\-v\fR.

.TP
\fB\-R [file]\fR
Remove file from the db and exit. Can be specified multiple
times. Duperemove will read the list from standard input if a hyphen
(-) is provided. Requires the \fB\--hashfile\fR option.

\fBNote:\fR If you are piping filenames from another duperemove instance it
is advisable to do so into a temporary file first as running duperemove
simultaneously on the same hashfile may corrupt that hashfile.

.TP
\fB\--fdupes\fR
Run in \fBfdupes\fR mode. With this option you can pipe the output of
\fBfdupes\fR to duperemove to dedupe any duplicate files found. When
receiving a file list in this manner, duperemove will skip the hashing phase.

.TP
\fB\-v\fR
Be verbose.

.TP
\fB\--skip-zeroes\fR
Read data blocks and skip any zeroed blocks, useful for speedup duperemove,
but can prevent deduplication of zeroed files.

.TP
\fB\-b size\fR
Use the specified block size for reading file extents. Defaults to 128K.

.TP
\fB\--io-threads=N\fR
Use N threads for I/O. This is used by the file hashing and dedupe
stages. Default is automatically detected based on number of
host cpus.

.TP
\fB\--cpu-threads=N\fR
Use N threads for CPU bound tasks. This is used by the duplicate
extent finding stage. Default is automatically detected based on
number of host cpus.

\fBNote:\fR Hyperthreading can adversely affect performance of the
extent finding stage. If duperemove detects an Intel CPU with
hyperthreading it will use half the number of cores reported by the
system for cpu bound tasks.

.TP
\fB\--dedupe-options=\fR\fIoptions\fR
Comma separated list of options which alter how we dedupe. Prepend 'no' to an
option in order to turn it off.
.RS
.TP
\fB[no]partial\fR
Duperemove can often find more dedupe by comparing portions of extents
to each other. This can be a lengthy, CPU intensive task so it is
turned off by default.

The code behind this option is under active development and as a
result the semantics of the \fIpartial\fR argument may change.

.TP
\fB[no]same\fR
Defaults to \fBoff\fR. Allow dedupe of extents within the same
file.
.TP
\fB[no]fiemap\fR
Defaults to \fBon\fR. Duperemove uses the \fIfiemap\fR ioctl during
the dedupe stage to optimize out already deduped extents as well as to
provide an estimate of the space saved after dedupe operations are
complete.

Unfortunately, some versions of Btrfs exhibit extremely poor
performance in fiemap as the number of references on a file extent
goes up. If you are experiencing the dedupe phase slowing down
or 'locking up' this option may give you a significant amount of
performance back.

\fBNote:\fR This does not turn off all usage of fiemap, to disable
fiemap during the file scan stage, you will also want to use the
\fB--lookup-extents=no\fR option.
.TP
\fB[no]block\fR
Deprecated.
.RE

.TP
\fB\--help\fR
Prints help text.

.TP
\fB\--lookup-extents=[yes|no]\fR
Defaults to \fByes\fR. Allows duperemove to skip checksumming some blocks by
checking their extent state.

.TP
\fB\--read-hashes=hashfile\fR
\fB\This option is primarily for testing.\fR See the \fB--hashfile\fR option if you want to use hashfiles.

Read hashes from a hashfile. A file list is not required with this
option. Dedupe can be done if duperemove is run from the same base
directory as is stored in the hash file (basically duperemove has to
be able to find the files).

.TP
\fB\--write-hashes=hashfile\fR
\fB\This option is primarily for testing.\fR See the \fB--hashfile\fR option if you want to use hashfiles.

Write hashes to a hashfile. These can be read in at a later date and
deduped from.

.TP
\fB\--debug\fR
Print debug messages, forces \fB-v\fR if selected.

.TP
\fB\--hash-threads=N\fR
Deprecated, see \fB--io-threads\fR above.

.TP
\fB\--hash=alg\fR
You can choose between murmur3 and xxhash. The default is murmur3 as
it is very fast and can generate 128 bit digests for a very small
chance of collision. Xxhash may be faster but generates only 64 bit
digests. Both hashes are fast enough that the default should work well
for the overwhelming majority of users.

.TP
\fB\--exclude=PATTERN\fR
You an exclude certain files and folders from the deduplication process. This
might be benefical for skipping subvolume snapshot mounts, for instance. You
need to provide full path for exclusion. For example providing just a file name
with a wildcard i.e \fBduperemove --exclude file-*\fR won't ever match because internally
duperemove works with absolute paths. Another thing to keep in mind is that
shells usually expand glob pattern so the passed in pattern ought to also be
quoted. Taking everything into consideration the correct way to pass an exclusion
pattern is \fBduperemove --exclude "/path/to/dir/file*" /path/to/dir\fR

.SH "EXAMPLES"
.SS "Simple Usage"
Dedupe the files in directory /foo, recurse into all subdirectories. You only want to use this for small data sets.
.IP
duperemove -dr /foo
.PP
Use duperemove with fdupes to dedupe identical files below directory foo.
.IP
fdupes -r /foo | duperemove --fdupes

.SS "Using Hashfiles"
Duperemove can optionally store the hashes it calculates in a
hashfile. Hashfiles have two primary advantages - memory usage and
re-usability. When using a hashfile, duperemove will stream computed
hashes to it, instead of main memory.

If Duperemove is run with an existing hashfile, it will only scan
those files which have changed since the last time the hashfile was
updated. The \fIfiles\fR argument controls which directories
duperemove will scan for newly added files. In the simplest usage, you
rerun duperemove with the same parameters and it will only scan
changed or newly added files - see the first example below.

.PP
Dedupe the files in directory foo, storing hashes in foo.hash. We can
run this command multiple times and duperemove will only checksum and
dedupe changed or newly added files.
.IP
duperemove -dr --hashfile=foo.hash foo/
.PP
Don't scan for new files, only update changed or deleted files, then dedupe.
.IP
duperemove -dr --hashfile=foo.hash
.PP
Add directory bar to our hashfile and discover any files that were
recently added to foo.
.IP
duperemove -dr --hashfile=foo.hash foo/ bar/
.PP
List the files tracked by foo.hash.
.IP
duperemove -L --hashfile=foo.hash

.SH "FAQ"
.SS Is there an upper limit to the amount of data duperemove can process?

Duperemove v0.11 is fast at reading and cataloging data. Dedupe runs will be
memory limited unless the '--hashfile' option is used. '--hashfile' allows
duperemove to temporarily store duplicated hashes to disk, thus removing the
large memory overhead and allowing for a far larger amount of data to be
scanned and deduped. Realistically though you will be limited by the speed of
your disks and cpu. In those situations where resources are limited you may
have success by breaking up the input data set into smaller pieces.

When using a hashfile, duperemove will only store duplicate hashes in
memory. During normal operation then the hash tree will make up the
largest portion of duperemove memory usage. As of Duperemove v0.11
hash entries are 88 bytes in size. If you know the number of duplicate
blocks in your data set you can get a rough approximation of memory
usage by multiplying with the hash entry size.

Actual performance numbers are dependent on hardware - up to date
testing information is kept on the duperemove wiki (see below for the link).

.SS How large of a hashfile will duperemove create?

Hashfiles are essentially sqlite3 database files with several tables,
the largest of which are the files and extents tables. Each extents
table entry is about 72 bytes though that may grow as features are
added. The size of a files table entry depends on the file path but a
good estimate is around 270 bytes per file. The number of extents in a
data set is directly proportional to file fragmentation level.

If you know the total number of extents and files in your data set then

duperemove.c
/*
 * duperemove.c
 *
 * Copyright (C) 2013 SUSE.  All rights reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License version 2 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * Authors: Mark Fasheh <mfasheh@suse.de>
 */

#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stddef.h>
#include <errno.h>
#include <string.h>
#include <getopt.h>
#include <inttypes.h>

#include <glib.h>

#include "list.h"
#include "csum.h"
#include "filerec.h"
#include "hash-tree.h"
#include "results-tree.h"
#include "dedupe.h"
#include "util.h"
#include "btrfs-util.h"
#include "dbfile.h"
#include "stats.h"
#include "memstats.h"
#include "debug.h"

#include "file_scan.h"
#include "find_dupes.h"
#include "run_dedupe.h"

#define MIN_BLOCKSIZE	(4U*1024)
/* max blocksize is somewhat arbitrary. */
#define MAX_BLOCKSIZE	(1024U*1024)
#define DEFAULT_BLOCKSIZE	(128U*1024)
unsigned int blocksize = DEFAULT_BLOCKSIZE;

int run_dedupe = 0;
int recurse_dirs = 0;
int v2_hashfile = 0;
int dedupe_same_file = 1;
int skip_zeroes = 0;

int target_rw = 1;
static int version_only = 0;
static int help_option = 0;
static int fdupes_mode = 0;
static int stdin_filelist = 0;
static unsigned int list_only_opt = 0;
static unsigned int rm_only_opt = 0;
struct dbfile_config dbfile_cfg;
static bool force_v2_hashfile = false;
static int partial_extent_search = 0;

static enum {
	H_READ,
	H_WRITE,
	H_UPDATE,
} use_hashfile = H_UPDATE;
char *serialize_fname = NULL;
static unsigned int nr_logical_cpus;
static unsigned int nr_physical_cpus;
unsigned int io_threads;
unsigned int cpu_threads;
int io_threads_opt = 0;
int cpu_threads_opt = 0;
int do_lookup_extents = 1;
int fiemap_during_dedupe = 1;

int stdout_is_tty = 0;

static char *user_hash = DEFAULT_HASH_STR;

static void print_file(char *filename, char *ino, char *subvol)
{
	if (verbose)
		printf("%s\t%s\t%s\n", filename, ino, subvol);
	else
		printf("%s\n", filename);
}

static int list_db_files(char *filename)
{
	int ret;

	ret = dbfile_open(filename, &dbfile_cfg);
	if (ret) {
		fprintf(stderr, "Error: Could not open \"%s\"\n", filename);
		return ret;
	}

	ret = dbfile_iter_files(dbfile_get_handle(), &print_file);

	dbfile_close();
	return ret;
}

struct rm_file {
	char *filename;
	struct list_head list;
};
static LIST_HEAD(rm_files_list);

static void add_rm_file(const char *filename)
{
	struct rm_file *rm = malloc(sizeof(*rm));
	if (rm) {
		rm->filename = strdup(filename);
		list_add_tail(&rm->list, &rm_files_list);
	}
}

static void free_rm_file(struct rm_file *rm)
{
	if (rm) {
		list_del(&rm->list);
		free(rm->filename);
		free(rm);
	}
}

static void add_rm_db_files_from_stdin(void)
{
	char *path = NULL;
	size_t pathlen = 0;
	ssize_t readlen;

	while ((readlen = getline(&path, &pathlen, stdin)) != -1) {
		if (readlen == 0)
			continue;

		if (readlen > 0 && path[readlen - 1] == '\n') {
			path[--readlen] = '\0';
		}

		if (readlen > PATH_MAX - 1) {
			fprintf(stderr, "Path max exceeded: %s\n", path);
			continue;
		}

		add_rm_file(path);
	}

	if (path != NULL)
		free(path);
}

static int rm_db_files(char *dbfilename)
{
	int ret, err = 0;
	struct rm_file *rm, *tmp;

	ret = dbfile_open(dbfilename, &dbfile_cfg);
	if (ret) {
		fprintf(stderr, "Error: Could not open \"%s\"\n", dbfilename);
		return ret;
	}

restart:
	list_for_each_entry_safe(rm, tmp, &rm_files_list, list) {
		if (strlen(rm->filename) == 1 && rm->filename[0] == '-') {
			add_rm_db_files_from_stdin();
			free_rm_file(rm);
			/*
			 * We may have added to the end of the list
			 * which messes up the next-entry condition
			 * for list_for_each_entry_safe()
			 */
			goto restart;
		}
		ret = dbfile_remove_file(dbfile_get_handle(), &dbfile_cfg,
					 rm->filename);
		if (ret == 0)
			vprintf("Removed \"%s\" from hashfile.\n",
				rm->filename);
		if (ret && ret != ENOENT && !err)
			err = ret;

		free_rm_file(rm);
	}

	dbfile_close();

	return err;
}

extern struct list_head exclude_list;

static void add_exclude_pattern(const char *pattern)
{
	struct exclude_file *exclude = malloc(sizeof(*exclude));
	if (exclude) {
		exclude->pattern = strdup(pattern);
		list_add_tail(&exclude->list, &exclude_list);
	}
}

static void usage(const char *prog)
{
	char *s = NULL;
#ifdef	DEBUG_BUILD
	s = " (debug build)";
#endif
	printf("duperemove %s%s\n", VERSTRING, s ? s : "");
	if (version_only)
		return;

	printf("Find duplicate extents and optionally dedupe them.\n\n");
	printf("Basic usage: %s [-r] [-d] [-h] [-v] [-A] "
	       "[--hashfile=hashfile] OBJECTS\n", prog);
	printf("\n\"OBJECTS\" is a list of files (or directories) which we\n");
	printf("want to find duplicate extents in. If a directory is \n");
	printf("specified, all regular files inside of it will be scanned.\n");
	printf("\n\t<switches>\n");
	printf("\t-r\t\tEnable recursive dir traversal.\n");
	printf("\t-d\t\tDe-dupe the results (must run on a supported fs).\n");
	printf("\t--hashfile=FILE\tStore hashes in this file.\n");
	printf("\t-A\t\tOpen files for dedupe in read-only mode.\n");
	printf("\t-h\t\tPrint numbers in human-readable format.\n");
	printf("\t-v\t\tPrint extra information (verbose).\n");
	printf("\t--help\t\tPrints this help text.\n");
	printf("\n\nPlease see the duperemove(8) manpage for a complete list "
	       "of options.\n");
}

static int parse_yesno_option(char *arg, int default_val)
{
	if (strncmp(arg, "yes", 3) == 0)
		return 1;
	else if (strncmp(arg, "no", 2) == 0)
		return 0;
	return default_val;
}

/* adapted from ocfs2-tools */
static int parse_dedupe_opts(const char *opts)
{
	char *options, *token, *next, *p, *arg;
	int print_usage = 0;
	int invert, ret = 0;

	options = strdup(opts);

	for (token = options; token && *token; token = next) {
		p = strchr(token, ',');
		next = NULL;
		invert = 0;

		if (p) {
			*p = '\0';
			next = p + 1;
		}

		arg = strstr(token, "no");
		if (arg == token) {
			invert = 1;
			token += strlen("no");
		}

		if (strcmp(token, "same") == 0) {
			dedupe_same_file = !invert;
		} else if (strcmp(token, "fiemap") == 0) {
			fiemap_during_dedupe = !invert;
		} else if (strcmp(token, "partial") == 0) {
			partial_extent_search = !invert;
		} else {
			print_usage = 1;
			break;
		}
	}

	if (print_usage) {
		fprintf(stderr, "Bad dedupe options specified. Valid dedupe "
			"options are:\n"
			"\t[no]same\n"
			"\t[no]fiemap\n"
			"\t[no]partial\n");
		ret = EINVAL;
	}

	free(options);
	return ret;
}

enum {
	DEBUG_OPTION = CHAR_MAX + 1,
	HELP_OPTION,
	VERSION_OPTION,
	WRITE_HASHES_OPTION,
	WRITE_OLD_HASHES_OPTION,
	READ_HASHES_OPTION,
	HASHFILE_OPTION,
	IO_THREADS_OPTION,
	CPU_THREADS_OPTION,
	LOOKUP_EXTENTS_OPTION,
	HASH_OPTION,
	SKIP_ZEROES_OPTION,
	FDUPES_OPTION,
	DEDUPE_OPTS_OPTION,
	QUIET_OPTION,
	EXCLUDE_OPTION,
};

static int add_files_from_stdin(int fdupes)
{
	int ret = 0;
	char *path = NULL;
	size_t pathlen = 0;
	ssize_t readlen;

	while ((readlen = getline(&path, &pathlen, stdin)) != -1) {
		if (readlen == 0)
			continue;

		if (fdupes && readlen == 1 && path[0] == '\n') {
			ret = fdupes_dedupe();
			if (ret)
				return ret;
			continue;
		}

		if (readlen > 0 && path[readlen - 1] == '\n') {
			path[--readlen] = '\0';
		}

		if (readlen > PATH_MAX - 1) {
			fprintf(stderr, "Path max exceeded: %s\n", path);
			continue;
		}

		if (add_file(path))
			return 1;

		/* Give the user a chance to see some output from add_file(). */
		if (!fdupes)
			fflush(stdout);
	}

	if (path != NULL)
		free(path);

	return 0;
}

static int add_files_from_cmdline(int numfiles, char **files)
{
	int i;

	for (i = 0; i < numfiles; i++) {
		const char *name = files[i];

		if (add_file(name))
			return 1;
	}

	return 0;
}

/*
 * Ok this is doing more than just parsing options.
 */
static int parse_options(int argc, char **argv, int *filelist_idx)
{
	int c, numfiles;
	int read_hashes = 0;
	int write_hashes = 0;
	int update_hashes = 0;

	static struct option long_ops[] = {
		{ "debug", 0, NULL, DEBUG_OPTION },
		{ "help", 0, NULL, HELP_OPTION },
		{ "version", 0, NULL, VERSION_OPTION },
		{ "write-hashes", 1, NULL, WRITE_HASHES_OPTION },
		{ "write-hashes-v2", 1, NULL, WRITE_OLD_HASHES_OPTION },
		{ "read-hashes", 1, NULL, READ_HASHES_OPTION },
		{ "hashfile", 1, NULL, HASHFILE_OPTION },
		{ "io-threads", 1, NULL, IO_THREADS_OPTION },
		{ "hash-threads", 1, NULL, IO_THREADS_OPTION },
		{ "cpu-threads", 1, NULL, CPU_THREADS_OPTION },
		{ "lookup-extents", 1, NULL, LOOKUP_EXTENTS_OPTION },
		{ "hash", 1, NULL, HASH_OPTION },
		{ "skip-zeroes", 0, NULL, SKIP_ZEROES_OPTION },
		{ "fdupes", 0, NULL, FDUPES_OPTION },
		{ "dedupe-options=", 1, NULL, DEDUPE_OPTS_OPTION },
		{ "quiet", 0, NULL, QUIET_OPTION },
		{ "exclude", 1, NULL, EXCLUDE_OPTION },
		{ NULL, 0, NULL, 0}
	};

	if (argc < 2)
		return 1;

	while ((c = getopt_long(argc, argv, "Ab:vdDrh?LR:q", long_ops, NULL))
	       != -1) {
		switch (c) {
		case 'A':
			target_rw = 0;
			break;
		case 'b':
			blocksize = parse_size(optarg);
			if (blocksize < MIN_BLOCKSIZE ||
			    blocksize > MAX_BLOCKSIZE){
				fprintf(stderr, "Error: Blocksize is bounded by %u and %u, %u found\n",
					MIN_BLOCKSIZE, MAX_BLOCKSIZE, blocksize);
				return EINVAL;
			}
			break;
		case 'd':
		case 'D':
			run_dedupe = 1;
			break;
		case 'r':
			recurse_dirs = 1;
			break;
		case VERSION_OPTION:
			version_only = 1;
			break;
		case DEBUG_OPTION:
			debug = 1;
			/* Fall through */
		case 'v':
			verbose = 1;
			break;
		case 'h':
			human_readable = 1;
			break;
		case WRITE_OLD_HASHES_OPTION:
			force_v2_hashfile = true;
		case WRITE_HASHES_OPTION:
			write_hashes = 1;
			serialize_fname = strdup(optarg);
			break;
		case READ_HASHES_OPTION:
			read_hashes = 1;
			serialize_fname = strdup(optarg);
			break;
		case HASHFILE_OPTION:
			update_hashes = 1;
			serialize_fname = strdup(optarg);
			break;
		case IO_THREADS_OPTION:
			io_threads = strtoul(optarg, NULL, 10);
			if (!io_threads){
				fprintf(stderr, "Error: --io-threads must be "
					"an integer, %s found\n", optarg);
				return EINVAL;
			}
			io_threads_opt = 1;
			break;
		case CPU_THREADS_OPTION:
			cpu_threads = strtoul(optarg, NULL, 10);
			if (!cpu_threads){
				fprintf(stderr, "Error: --cpu-threads must be "
					"an integer, %s found\n", optarg);
				return EINVAL;
			}
			cpu_threads_opt = 1;
			break;
		case LOOKUP_EXTENTS_OPTION:
			do_lookup_extents = parse_yesno_option(optarg, 0);
			break;
		case HASH_OPTION:
			user_hash = optarg;
			break;
		case SKIP_ZEROES_OPTION:
			skip_zeroes = 1;
			break;
		case FDUPES_OPTION:
			fdupes_mode = 1;
			break;
		case DEDUPE_OPTS_OPTION:
			if (parse_dedupe_opts(optarg))
				return EINVAL;
			break;
		case 'L':
			list_only_opt = 1;
			break;
		case 'R':
			rm_only_opt = 1;
			add_rm_file(optarg);
			break;
		case QUIET_OPTION:
		case 'q':
			quiet = 1;
			break;
		case EXCLUDE_OPTION:
			add_exclude_pattern(optarg);
			break;
		case HELP_OPTION:
			help_option = 1;
		case '?':
		default:
			version_only = 0;
			return 1;
		}
	}

	if (!do_lookup_extents || !fiemap_during_dedupe)
		force_v2_hashfile = true;

	numfiles = argc - optind;

	/* Filter out option combinations that don't make sense. */
	if ((write_hashes + read_hashes + update_hashes) > 1) {
		fprintf(stderr, "Error: Specify only one hashfile option.\n");
		return 1;
	}

	if (read_hashes)
		use_hashfile = H_READ;
	else if (write_hashes)
		use_hashfile = H_WRITE;
	else if (update_hashes)
		use_hashfile = H_UPDATE;

	if (read_hashes) {
		if (numfiles) {
			fprintf(stderr,
				"Error: --read-hashes option does not take a "
				"file list argument\n");
			return 1;
		}
		goto out_nofiles;
	}

	if (fdupes_mode) {
		if (read_hashes || write_hashes || update_hashes) {
			fprintf(stderr,
				"Error: cannot mix hashfile option with "
				"--fdupes option\n");
			return 1;
		}

		if (numfiles) {
			fprintf(stderr,
				"Error: fdupes option does not take a file "
				"list argument\n");
			return 1;
		}
		/* rest of fdupes mode is implemented in main() */
		return 0;
	}

	*filelist_idx = 0;
	if (numfiles == 1 && strcmp(argv[optind], "-") == 0)
		stdin_filelist = 1;
	else {
		*filelist_idx = optind;
	}

	if (list_only_opt && rm_only_opt) {
		fprintf(stderr, "Error: Can not mix '-L' and '-R' options.\n");
		return 1;
	}

	if (list_only_opt || rm_only_opt) {
		if (!serialize_fname || use_hashfile == H_WRITE) {
			fprintf(stderr,	"Error: --hashfile= option is required "
				"with '-L' or -R.\n");
			return 1;
		}

		if (numfiles) {
			fprintf(stderr, "Error: -L and -R options do not take "
				"a file list argument\n");
			return 1;
		}
	}

out_nofiles:

	return 0;
}

static void print_header(void)
{
	vprintf("Using %uK blocks\n", blocksize / 1024);
	vprintf("Using hash: %s\n", csum_mod->name);
	vprintf("Using %s hashing\n", v2_hashfile ? "block-based" : "extent-based");
#ifdef	DEBUG_BUILD
	printf("Debug build, performance may be impacted.\n");
#endif
	qprintf("Gathering file list...\n");
}

static int update_config_from_dbfile(void)
{
	dedupe_seq = dbfile_cfg.dedupe_seq;

	if (strncasecmp(dbfile_cfg.hash_type, hash_type, 8)) {
		fprintf(stderr,
			"Error: Hashfile %s uses %.*s. for checksums "
			"but we are using %.*s.\nTry running with "
			"--hash=%.*s\n", serialize_fname, 8,
			dbfile_cfg.hash_type, 8, hash_type,
			8, dbfile_cfg.hash_type);
		return EINVAL;
	}

	if (dbfile_cfg.blocksize != blocksize &&
	    dbfile_cfg.major == BLOCK_DEDUPE_DBFILE_VER) {
		vprintf("Using blocksize %uK from hashfile (%uK "
			"blocksize requested).\n", dbfile_cfg.blocksize/1024,
			blocksize/1024);
		blocksize = dbfile_cfg.blocksize;
	}

	return 0;
}

static int create_update_hashfile(int argc, char **argv, int filelist_idx)
{
	int ret;
	int dbfile_is_new = 0;

	ret = dbfile_create(serialize_fname, &dbfile_is_new,
			    force_v2_hashfile ? BLOCK_DEDUPE_DBFILE_VER : DB_FILE_MAJOR,
			    &dbfile_cfg);
	if (ret)
		goto out;

	if (force_v2_hashfile && dbfile_cfg.major != BLOCK_DEDUPE_DBFILE_VER) {
		ret = EINVAL;
		fprintf(stderr, "Error: asked to force hashfile version 2 but "
			"existing hashfile has version %d\n", dbfile_cfg.major);
		goto out;
	}

	if (dbfile_cfg.major == BLOCK_DEDUPE_DBFILE_VER)
		v2_hashfile = 1;

	if (!dbfile_is_new) {
		ret = update_config_from_dbfile();
		if (ret)
			goto out;
		fs_set_onefs(dbfile_cfg.onefs_dev, dbfile_cfg.onefs_fsid);
	}

	print_header();

	if (stdin_filelist)
		ret = add_files_from_stdin(0);
	else
		ret = add_files_from_cmdline(argc - filelist_idx,
					     &argv[filelist_idx]);
	if (ret)
		goto out;

	if (dbfile_is_new) {
		dbfile_cfg.blocksize = blocksize;
		dbfile_cfg.onefs_dev = fs_onefs_dev();
		dbfile_cfg.onefs_fsid = fs_onefs_id();
		dbfile_cfg.dedupe_seq = dedupe_seq;
		ret = dbfile_sync_config(&dbfile_cfg);
		if (ret)
			goto out;
	} else {
		qprintf("Adding files from database for hashing.\n");

		ret = dbfile_scan_files(&dbfile_cfg);
		if (ret)
			goto out;
	}

	if (list_empty(&filerec_list)) {
		fprintf(stderr, "No dedupe candidates found.\n");
		ret = EINVAL;
		goto out;
	}

	ret = populate_tree(&dbfile_cfg);
	if (ret) {
		fprintf(stderr,	"Error while populating extent tree!\n");
		goto out;
	}

	ret = create_indexes(dbfile_get_handle(), &dbfile_cfg);
	if (ret)
		goto out;

	/*
	 * File scan from above can cause quite a bit of output, flush
	 * here in case of logfile.
	 */
	if (stdout_is_tty)
		fflush(stdout);

	ret = dbfile_sync_files(dbfile_get_handle());
	if (ret)
		goto out;
out:
	return ret;
}

int main(int argc, char **argv)
{
	int ret, filelist_idx = 0;
	struct results_tree res;
	struct hash_tree dups_tree;

	init_filerec();
	init_results_tree(&res);
	init_hash_tree(&dups_tree);

	if (!geteuid())
		target_rw = 0;
	ret = parse_options(argc, argv, &filelist_idx);
	if (ret || version_only) {
		usage(argv[0]);
		return (version_only || help_option) ? 0 : EINVAL;
	}

	/* Allow larger than unusal amount of open files. On linux
	 * this should bw increase form 1K to 512K open files
	 * simultaneously.
	 *
	 * On multicore SSD machines it's not hard to get to 1K open
	 * files.
	 */
	increase_limits();
	/*
	 * Don't run detection if the user has supplied our cpu counts
	 * already.
	 */
	if (!io_threads_opt || !cpu_threads_opt) {
		get_num_cpus(&nr_physical_cpus, &nr_logical_cpus);
		if (!io_threads)
			io_threads = nr_logical_cpus;
		if (!cpu_threads)
			cpu_threads = nr_physical_cpus;
	}

	if (fdupes_mode)
		return add_files_from_stdin(1);

	ret = init_csum_module(user_hash);
	if (ret) {
		if (ret == EINVAL)
			fprintf(stderr,
				"Could not initialize hash module \"%s\"\n",
				user_hash);
		return ret;
	}

	if (isatty(STDOUT_FILENO))
		stdout_is_tty = 1;

	if (list_only_opt)
		return list_db_files(serialize_fname);
	else if (rm_only_opt)
		return rm_db_files(serialize_fname);

	switch (use_hashfile) {
	case H_UPDATE:
	case H_WRITE:
		if (serialize_fname)
			printf("Warning: The hash file format in Duperemove "
			       "master branch is under development and may "
			       "change.\nIf the changes are not backwards "
			       "compatible, you will have to re-create your "
			       "hash file.\n");
		ret = create_update_hashfile(argc, argv, filelist_idx);
		if (ret)
			goto out;

		if (use_hashfile == H_WRITE) {
			/*
			 * This option is for isolating the file scan
			 * stage. Exit the program now.
			 */
			qprintf("Hashfile \"%s\" written, exiting.\n",
				serialize_fname);
			goto out;
		}
		break;
	case H_READ:
		ret = dbfile_open(serialize_fname, &dbfile_cfg);
		if (ret) {
			fprintf(stderr, "Error: Could not open dbfile %s.\n",
				serialize_fname);
			goto out;
		}

		if (dbfile_cfg.major == BLOCK_DEDUPE_DBFILE_VER)
			v2_hashfile = 1;

		/*
		 * Skips the file scan, used to isolate the
		 * extent-find and dedupe stages
		 */
		blocksize = dbfile_cfg.blocksize;
		ret = update_config_from_dbfile();
		if (ret)
			goto out;

		print_header();
		break;
	default:
		abort_lineno();
		break;
	}

	qprintf("Loading only duplicated hashes from hashfile.\n");

	if (v2_hashfile) {
		ret = dbfile_load_block_hashes(&dups_tree);
		if (ret)
			goto out;

		ret = find_all_dupes(&dups_tree, &res);
		if (ret) {
			/* Only error for this should be enomem */
			fprintf(stderr,
				"Error %d: %s while finding duplicate extents.\n",
				ret, strerror(ret));
			goto out;
		}
	} else {
		ret = dbfile_load_extent_hashes(&res);
		if (ret)
			goto out;

		ret = dbfile_load_block_hashes(&dups_tree);
		if (ret)
			goto out;

		printf("Found %llu identical extents.\n", res.num_extents);
		if (partial_extent_search) {
			ret = find_additional_dedupe(&dups_tree, &res);
			if (ret)
				goto out;
		}
	}

#ifdef	PRINT_STATS
	run_filerec_stats();
#endif

	if (run_dedupe) {
		dedupe_results(&res, &dups_tree);

		/*
		 * Bump dedupe_seq, this effectively marks the files
		 * in our hashfile as having been through dedupe.
		 */
		dedupe_seq++;

		/* Sync to get new dedupe_seq written. */
		dbfile_cfg.dedupe_seq = dedupe_seq;
		dbfile_cfg.blocksize = blocksize;
		dbfile_cfg.onefs_dev = fs_onefs_dev();
		dbfile_cfg.onefs_fsid = fs_onefs_id();
		ret = dbfile_sync_config(&dbfile_cfg);
		if (ret)
			goto out;
	} else {
		print_dupes_table(&res);
	}

out:
	free_results_tree(&res);
	free_hash_tree(&dups_tree);
	free_all_filerecs();
	dbfile_close();

#ifdef DEBUG_BUILD
	print_mem_stats();
#else
	if (ret == ENOMEM || debug)
		print_mem_stats();
#endif

	return ret;
}
